Corona-Warn-App/src/main/java/de/rki/coronawarnapp/task/TaskController.kt

```diff
@@ -0,0 +1,219 @@
+package de.rki.coronawarnapp.task
+
+import androidx.annotation.VisibleForTesting
+import de.rki.coronawarnapp.task.TaskFactory.Config.CollisionBehavior
+import de.rki.coronawarnapp.util.TimeStamper
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.CoroutineStart
+import kotlinx.coroutines.async
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.launchIn
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.flow.onCompletion
+import kotlinx.coroutines.flow.onEach
+import kotlinx.coroutines.flow.onStart
+import kotlinx.coroutines.flow.receiveAsFlow
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+import timber.log.Timber
+import java.util.UUID
+import javax.inject.Inject
+import javax.inject.Singleton
+
+@Singleton
+class TaskController @Inject constructor(
+    private val taskFactories: Map<
+        @JvmSuppressWildcards Class<out Task<*, *>>,
+        @JvmSuppressWildcards TaskFactory<out Task.Progress, out Task.Result>>,
+    @TaskCoroutineScope private val taskScope: CoroutineScope,
+    private val timeStamper: TimeStamper
+) {
+
+    private val mutex = Mutex()
+    private val taskQueue = Channel<TaskRequest>(Channel.UNLIMITED)
+    private val internalTaskData = MutableStateFlow<Map<UUID, InternalTaskState>>(
+        emptyMap()
+    )
+
+    private val submissionProcessor = taskQueue
+        .receiveAsFlow()
+        .onStart { Timber.tag(TAG).v("Listening to task request queue.") }
+        .onEach { initTaskData(it) }
+        .onCompletion { Timber.tag(TAG).w("Stopped listening to request queue. Why?") }
+        .launchIn(taskScope)
+
+    val tasks: Flow<List<TaskInfo>> = internalTaskData
+        .map { it.values }
+        .map { tasks ->
+            tasks.map {
+                TaskInfo(it, it.task.progress)
+            }
+        }
+
+    init {
+        Timber.tag(TAG).d("We have factories for %s", taskFactories.keys)
+    }
+
+    /**
+     * Don't re-use taskrequests, create new requests for each submission.
+     * They contain unique IDs.
+     */
+    fun submit(request: TaskRequest) {
+        if (!taskFactories.containsKey(request.type.java)) {
+            throw MissingTaskFactoryException(request::class)
+        }
+        Timber.tag(TAG).i("Task submitted: %s", request)
```

Comment:
This log only holds true after the ".offer" completed with true. It could also fail

Reply:
offer will always return `true` here.
We are using a `Channel(Channel.UNLIMITED)` as queue.

```
When capacity is Channel.UNLIMITED — it creates a LinkedListChannel. This channel has a linked-list buffer of unlimited capacity (limited only by available memory). Sending to this channel never suspends, and offer always returns true.
```